spectator-reg-atlas/src/main/java/com/netflix/spectator/atlas/impl/PrefixTree.java

```diff
@@ -0,0 +1,206 @@
+ * Simple tree for finding all values associated with a prefix that matches the search
+ * key. The prefix is a simple ascii string. If unsupported characters are used in the
+ * prefix or search key, then the prefix match will only check up to the unsupported
+ * character and the caller will need to perform further checks on the returned value.
+ */
+final class PrefixTree<T> {
+
+  private static final int FIRST_CHAR = ' ';
+  private static final int LAST_CHAR = '~';
+  private static final int TABLE_SIZE = LAST_CHAR - FIRST_CHAR + 1;
+
+  private static int indexOf(char c) {
+    int i = c - FIRST_CHAR;
+    return (i >= TABLE_SIZE) ? -1 : i;
+  }
+
+  private final AtomicReferenceArray<PrefixTree<T>> children;
+  private final Set<T> values;
+
+  /** Create a new instance. */
+  PrefixTree() {
+    children = new AtomicReferenceArray<>(TABLE_SIZE);
+    values = ConcurrentHashMap.newKeySet();
+  }
+
+  private PrefixTree<T> computeIfAbsent(int i) {
+    PrefixTree<T> child = children.get(i);
+    if (child == null) {
+      synchronized (this) {
+        child = children.get(i);
+        if (child == null) {
+          child = new PrefixTree<>();
+          children.set(i, child);
+        }
+      }
+    }
+    return child;
+  }
+
+  /**
+   * Put a value into the tree.
+   *
+   * @param prefix
+   *     ASCII string that represents a prefix for the search key.
+   * @param value
+   *     Value to associate with the prefix.
+   */
+  void put(String prefix, T value) {
+    if (prefix == null)
+      values.add(value);
+    else
+      put(prefix, 0, value);
+  }
+
+  private void put(String prefix, int pos, T value) {
+    if (pos == prefix.length()) {
+      values.add(value);
+    } else {
+      int i = indexOf(prefix.charAt(pos));
+      if (i < 0) {
+        values.add(value);
+      } else {
+        PrefixTree<T> child = computeIfAbsent(i);
+        child.put(prefix, pos + 1, value);
+      }
+    }
+  }
+
+  /**
+   * Remove a value from the tree with the associated prefix.
+   *
+   * @param prefix
+   *     ASCII string that represents a prefix for the search key.
+   * @param value
+   *     Value to associate with the prefix.
+   * @return
+   *     Returns true if a value was removed from the tree.
+   */
+  boolean remove(String prefix, T value) {
+    if (prefix == null)
+      return values.remove(value);
+    else
+      return remove(prefix, 0, value);
+  }
+
+  private boolean remove(String prefix, int pos, T value) {
+    if (pos == prefix.length()) {
+      return values.remove(value);
+    } else {
+      int i = indexOf(prefix.charAt(pos));
+      if (i < 0) {
+        return values.remove(value);
+      } else {
+        PrefixTree<T> child = children.get(i);
+        if (child == null) {
+          return false;
+        } else {
+          boolean result = child.remove(prefix, pos + 1, value);
+          if (result && child.isEmpty()) {
+            synchronized (this) {
+              if (child.isEmpty()) {
```

Comment:
Need another `children.get(i);` to double check that the child hasn't been emptied, nulled, then replaced between L120 and L125.